### PR TITLE
Remove a href=# elements and replace with buttons

### PR DESCRIPTION
**Pull request by ajpal**
@dsjen @thallada 
@edx/edx-accessibility

### CODE REVIEW

**Comment by thallada** [comment]
:100:

File: acceptance_tests/mixins.py [hunk 1]
```diff
@@ -156,7 +156,7 @@ def _test_user_menu(self):
         """
         Verify the user menu functions properly.
         """
-        element = self.page.q(css='a.active-user.dropdown-toggle')
+        element = self.page.q(css='button.active-user.dropdown-toggle')
```
Comment (dsjen) [inline acceptance_tests/mixins.py]: 😸

File: analytics_dashboard/static/apps/learners/roster/views/search.js [hunk 2]
```diff
@@ -69,6 +73,7 @@ define(function(require) {
         clear: function(event) {
             event.preventDefault();
             this.collection.unsetSearchString();
+            this.searchBox().val('');
```
Comment (dsjen) [inline analytics_dashboard/static/apps/learners/roster/views/search.js:80]: Thanks for fixing that!

File: analytics_dashboard/static/sass/style-application.scss [hunk 3]
```diff
@@ -16,7 +16,7 @@
 @import "../bower_components/font-awesome/scss/font-awesome";
 @import "../vendor/dataTables/dataTables.bootstrap";
 @import "../vendor/dataTables/dataTables.fontAwesome";
-@import "../bower_components/nvd3/build/nv.d3";
+@import "../bower_components/nvd3/build/nv.d3.css";
```
Comment (dsjen) [inline analytics_dashboard/static/sass/style-application.scss]: I believe that the nv.d3.scss is created through a bower script.  Is this change needed for some reason?

Comment (ajpal) [inline analytics_dashboard/static/sass/style-application.scss]: Without this change, I got this error when I ran make static:
Error: It's not clear which file to import for '@import "../bower_components/nvd3/build/nv.d3"
Candidates:
         ../bower_components/nvd3/build/nv.d3.scss
         ../bower_components/nvd3/build/nv.d3.css
       Please delete or rename all but one of these files.

File: analytics_dashboard/static/sass/_developer.scss [hunk 4]
```diff
@@ -22,12 +22,49 @@
...
 }
 
+.lens-nav .navbar-nav .dropdown button {
+  color: $navbar-default-link-active-color;
+  text-transform: uppercase;
+  background: transparent
+}
+
+.lens-nav .navbar-nav li > button {
+  min-height: 55px;
```
Comment (dsjen) [inline analytics_dashboard/static/sass/_developer.scss]: I believe this height comes from `$lens-nav-height`.  Actually, I think that a lot of these navbar improvements could be moved into and replace the existing styles, https://github.com/edx/edx-analytics-dashboard/blob/f6bc937de58852f56546326d10f6a96a262ae0c6/analytics_dashboard/static/sass/_developer.scss#L203.  What do you think?

File: analytics_dashboard/static/sass/style-application.scss (hunk 3 above)
Comment (dsjen) [inline analytics_dashboard/static/sass/style-application.scss]: Oh!  I think you'll need to delete the nvd3 bower directory and run `bower install` again.  I updated the bower script to "move" the `.css` to `.scss` instead, but you probably still have both of them there.  Please let me know if that works for you!

Comment (ajpal) [inline analytics_dashboard/static/sass/style-application.scss]: I'm still getting the same error...
It's weird, though, because it looks like the nv.d3.css file is no longer in the directory. But it's still giving an error about it

Comment (ajpal) [inline analytics_dashboard/static/sass/style-application.scss]: finally fixed :)

Comment (dsjen) [inline analytics_dashboard/static/sass/style-application.scss]: 🙌  🙌  🙌

File: analytics_dashboard/static/apps/learners/roster/views/search.js [hunk 5]
```diff
@@ -37,6 +41,10 @@ define(function(require) {
             Backgrid.Extension.ServerSideFilter.prototype.initialize.call(this, options);
         },
 
+        clearButton: function() {
```
Comment (ajpal) [inline analytics_dashboard/static/apps/learners/roster/views/search.js:44]: This allows `showClearButtonMaybe()` to work correctly, but should be deleted once Backgrid's `clearButton` function is fixed to expect a `button` element instead of an `a` element

File: acceptance_tests/mixins.py (hunk 1 above)
Comment (downzer0) [inline acceptance_tests/mixins.py]: I would omit the element altogether.

File: analytics_dashboard/courses/templates/courses/home.html [hunk 6]
```diff
@@ -11,7 +11,7 @@
     <div class="container ">
       <ul class="nav navbar-nav">
         <li class="active nav-section">
-          <a href="#"><span class="link-label">{% trans "Course Home" %}</span></a>
+          <div class="course-label">{% trans "Course Home" %}</div>
```
Comment (downzer0) [inline analytics_dashboard/courses/templates/courses/home.html:14]: You went from an `a` to a `div`... is this meant to be clickable?

File: analytics_dashboard/static/apps/learners/roster/views/search.js [hunk 7]
```diff
@@ -37,6 +41,10 @@ define(function(require) {
...
+            return this.$el.find('button[data-backgrid-action=clear]');
```
Comment (downzer0) [inline analytics_dashboard/static/apps/learners/roster/views/search.js]: I would omit the element here too, so just `[data-backgrid-action="clear"]`

File: analytics_dashboard/static/apps/learners/roster/views/spec/roster-spec.js [hunk 8]
```diff
@@ -560,7 +560,7 @@ define(function(require) {
                 executeSearch(searchString);
                 expectSearchedFor(searchString);
                 getLastRequest().respond(200, {}, JSON.stringify(getResponseBody(1, 1)));
-                $('a.clear').click();
+                $('button.clear').click();
```
Comment (downzer0) [inline analytics_dashboard/static/apps/learners/roster/views/spec/roster-spec.js]: Can you use just `.clear`, without the element?

File: analytics_dashboard/templates/header.html [hunk 9]
```diff
@@ -34,12 +34,12 @@
           </li>
         {% endif %}
         <li class="dropdown">
-          <a href="#" class="dropdown-toggle active-user navbar-link" data-toggle="dropdown" role="button"
+          <button class="dropdown-toggle active-user navbar-link" data-toggle="dropdown" role="button"
```
Comment (downzer0) [inline analytics_dashboard/templates/header.html]: The `role` here is unnecessary as you're already using a `button`.

File: analytics_dashboard/templates/header.html [hunk 10]
```diff
@@ -34,12 +34,12 @@
...
              id="userMenu">
             <span class="ico-user fa fa-graduation-cap" aria-hidden="true"></span>
             <span class="active-user-name">{{ request.user.username }}</span>
             <span class="ico ico-indicator fa fa-caret-down" aria-hidden="true"></span>
-          </a>
+          </button>
 
           <ul class="dropdown-menu active-user-nav" role="menu" aria-labelledby="userMenu">
```
Comment (downzer0) [inline analytics_dashboard/templates/header.html]: This should probably have a `tabindex="-1"` so it's programmatically focusable when the menu is shown.

Comment (downzer0) [inline analytics_dashboard/templates/header.html]: I would also remove the `role="menu"` here as this isn't using Menu correctly.

File: analytics_dashboard/templates/header.html (hunk 9 above)
Comment (downzer0) [inline analytics_dashboard/templates/header.html]: Add `aria-controls="{id of menu it opens}"` that references the unique ID of the dropdown menu this toggles. You'll want to add a unique ID to the menu below.

File: analytics_dashboard/templates/header.html (hunk 10 above)
Comment (downzer0) [inline analytics_dashboard/templates/header.html]: Let's also add a unique ID here that we can then reference with `aria-controls` in the toggle button.

File: analytics_dashboard/templates/lens-navigation.html [hunk 11]
```diff
@@ -9,14 +9,14 @@
     <div class="container ">
       <ul class="nav navbar-nav">
         <li class="dropdown">
-          <a href="#" class="dropdown-toggle navbar-link" data-toggle="dropdown" role="button" id="lensPrimaryMenu">
+          <button class="dropdown-toggle navbar-link" data-toggle="dropdown" role="button" id="lensPrimaryMenu">
```
Comment (downzer0) [inline analytics_dashboard/templates/lens-navigation.html]: Let's remove the `role` as it's unnecessary. Let's also give this guy an `aria-controls=""` which references a unique ID of the menu it toggles.

File: analytics_dashboard/templates/lens-navigation.html [hunk 12]
```diff
@@ -9,14 +9,14 @@
...
             <span class="link-label">
             <span class="ico fa {{ primary_nav_item.icon }}" aria-hidden="true"></span>
             {{ primary_nav_item.label }}
             {# Translators: This refers to the active tab/navigation item. #}
             <span class="sr-only">({% trans "Active" %})</span>
             <span class="ico ico-indicator fa fa-caret-down" aria-hidden="true"></span></span>
-          </a>
+          </button>
 
           <ul class="dropdown-menu" role="menu" aria-labelledby="lensPrimaryMenu">
```
Comment (downzer0) [inline analytics_dashboard/templates/lens-navigation.html]: Let's remove the `role` as we aren't implementing the correct use of Menu. We should also make this focusable with `tabindex="-1"` and add a unique ID that is referenced by the toggle button in `aria-controls`.

**Comment by downzer0** [comment]
Looking good, @ajpal. I left a few comments pertaining to the functionality of the dropdown menus and their toggle buttons.

We aren't using the correct implementation of an actual "menu" so using the menu role actually makes this more confusing to use for screenreader users. Using `role="menu"` would be creating a custom widget, and we'd need to use `role="menuitem"` on its children and create additional keyboard functionality that would simulate an actual OS menu.

When a toggle button opens a dropdown menu, we should send focus to the menu (hence the `tabindex="-1"`), specifically the first menu item in the menu. Ideally, one could move between menu items with the arrow keys, only able to exit the menu with the Esc key, in which focus is sent back to the toggle button.

Finally, I think it's best to target only classnames or some sort of attribute in code (JS and CSS) as it allows for better flexibility down the road. For example, we could target `button.some-class`, but targeting `.some-class` instead accomplishes the same goal, but allows us to use whatever element we may need without having to update tests. The same goes for CSS - targeting classes rather than elements - but knowing our platform, this is probably way more time consuming and probably best left for another day.

File: analytics_dashboard/courses/templates/courses/home.html (hunk 6 above)
Comment (ajpal) [inline analytics_dashboard/courses/templates/courses/home.html:14]: No, clicking it didn't actually do anything. My guess is that it was an `a` element just to mimic the style of the rest of the links/buttons in the navbar

**Comment by ajpal** [comment]
Addressed comments. Let me know if there are any other changes that should be made 😄 
@dsjen @thallada @clrux

File: analytics_dashboard/static/apps/learners/roster/views/search.js (hunk 5 above)
Comment (dsjen) [reply analytics_dashboard/static/apps/learners/roster/views/search.js:44]: Thank you!  Would you mind adding a comment about that in the code so that we don't lose this?

File: analytics_dashboard/static/sass/_developer.scss [hunk 13]
```diff
@@ -155,6 +173,16 @@ button.chart-info {
...
   }
+
+  li > button {
+    line-height: 20px;
+    border: none;
+    background: transparent;
+  }
+
+  li > button:focus {
+    outline: dotted 1px #000;
```
Comment (dsjen) [inline analytics_dashboard/static/sass/_developer.scss:175]: 💠

File: analytics_dashboard/static/sass/_developer.scss [hunk 14]
```diff
@@ -22,12 +22,30 @@
...
+.clear.btn {
+  background: transparent;
+}
+
+.course-label {
+  padding-top: 18px;
+  padding-bottom: 0px;
+  padding-left: 15px;
+  padding-right: 15px;
+  min-height: 55px;
```
Comment (dsjen) [inline analytics_dashboard/static/sass/_developer.scss]: I think we can use `$lens-nav-height` here.

File: analytics_dashboard/static/sass/_developer.scss [hunk 15]
```diff
@@ -22,12 +22,30 @@
...
 
+.clear.btn {
+  background: transparent;
+}
+
+.course-label {
+  padding-top: 18px;
+  padding-bottom: 0px;
+  padding-left: 15px;
+  padding-right: 15px;
```
Comment (dsjen) [inline analytics_dashboard/static/sass/_developer.scss]: I believe that `15px` in the paddings are the same as what is in https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/static/sass/_developer.scss#L177.  Pulling this into a sass variable would be more maintainable.

File: analytics_dashboard/static/sass/_developer.scss [hunk 16]
```diff
@@ -22,12 +22,30 @@
...
   padding-top: 0;
   border: none;
 }
 
+.clear.btn {
+  background: transparent;
+}
+
+.course-label {
+  padding-top: 18px;
```
Comment (dsjen) [inline analytics_dashboard/static/sass/_developer.scss]: The `18px` is the same as https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/static/sass/_developer.scss#L205.  If we can pull this into a sass variable, it'll help with maintainability.

**Comment by dsjen** [comment]
@ajpal thank you so much for taking on this.  I left a few comments about pulling some magic numbers into sass variables.  Otherwise, 👍 !

Great job!

**Comment by dsjen** [comment]
Thanks for the lovely css refactor!  🏆

File: analytics_dashboard/templates/header.html [hunk 17]
```diff
@@ -34,14 +34,15 @@
...
+             aria-controls="userDropdownMenu" id="userMenu">
             <span class="ico-user fa fa-graduation-cap" aria-hidden="true"></span>
             <span class="active-user-name">{{ request.user.username }}</span>
             <span class="ico ico-indicator fa fa-caret-down" aria-hidden="true"></span>
-          </a>
+          </button>
 
-          <ul class="dropdown-menu active-user-nav" role="menu" aria-labelledby="userMenu">
+          <ul id="userDropdownMenu" class="dropdown-menu active-user-nav" tabindex="-1"
+              aria-labelledby="userMenu">
```
Comment (downzer0) [inline analytics_dashboard/templates/header.html:46]: This is referencing the `id` of the button and the accessible name will read "username". Since this is a menu for the username, we may want to indicate as such. You can use `sr` text for this like so:

```
...
<span class="active-user-name"><span class="sr">Menu for </span>{{ request.user.username }}</span>
...
```

This way the dropdown menu will have the accessible name of "Menu for username", and provide a bit more context.

Comment (downzer0) [inline analytics_dashboard/templates/header.html:46]: Obviously you'll want to translate the additional text.

File: analytics_dashboard/templates/lens-navigation.html [hunk 18]
```diff
@@ -9,16 +9,17 @@
...
             <span class="ico fa {{ primary_nav_item.icon }}" aria-hidden="true"></span>
             {{ primary_nav_item.label }}
             {# Translators: This refers to the active tab/navigation item. #}
             <span class="sr-only">({% trans "Active" %})</span>
             <span class="ico ico-indicator fa fa-caret-down" aria-hidden="true"></span></span>
-          </a>
+          </button>
 
-          <ul class="dropdown-menu" role="menu" aria-labelledby="lensPrimaryMenu">
+          <ul id="lensDropdownMenu" class="dropdown-menu" aria-labelledby="lensPrimaryMenu" tabindex="-1">
```
Comment (downzer0) [inline analytics_dashboard/templates/lens-navigation.html:23]: The same thing here applies. Currently, this accessible name will read "Label Active", but we're lacking the **what**. We are in a menu for something, so let's add `sr` text that indicates as much. See the above comment, and make sure it's translated properly.

File: analytics_dashboard/templates/lens-navigation.html [hunk 19]
```diff
@@ -12,6 +12,7 @@
           <button class="dropdown-toggle navbar-link" data-toggle="dropdown" id="lensPrimaryMenu"
               aria-controls="lensDropdownMenu">
             <span class="link-label">
+            <span class="sr-only">{% trans "Lens Menu." %}</span>
```
Comment (downzer0) [inline analytics_dashboard/templates/lens-navigation.html]: Do we need the period?

Comment (downzer0) [inline analytics_dashboard/templates/lens-navigation.html]: And is "lens" clear? I think of glass :)

Comment (ajpal) [reply analytics_dashboard/templates/lens-navigation.html]: When I was testing, it said "Lens Menu Engagement Active" all at once, and I thought it was more clear with a small pause: "Lens Menu (pause) Engagement Active" which is why I added the period.
Throughout Insights, we refer to the different components (Enrollment, Engagement, Performance, and Learners) as lenses, so I think "Lens Menu" is the most clear name for the button.

Comment (dsjen) [reply analytics_dashboard/templates/lens-navigation.html]: We can add a translator comment, e.g. `{# Translators: Lens is a collection of pages in our application. #}` -- example in https://github.com/edx/edx-analytics-dashboard/blob/cdce7a5aead9cd65f8213c6848f6cfbcfbc89fef/analytics_dashboard/courses/templates/courses/views_chart_info.html#L3

Comment (downzer0) [reply analytics_dashboard/templates/lens-navigation.html]: It's less about the translation, and more about will non-sighted users, or users with cognitive impairments understand what a "lens" is. So, this menu is a collection of pages, similar to navigation? If so, can we call it "Navigation menu"?

Comment (thallada) [reply analytics_dashboard/templates/lens-navigation.html]: I personally dislike the name "lens" because it's unclear (I think it's fine to use it internally in the code). It doesn't look like we display that word anywhere in the UI as well, so I think it makes sense to use "navigation" in the screen reader text here.

**Comment by dsjen** [comment]
The weird flaky a11y error happened again!  I restarted the build.

File: analytics_dashboard/templates/lens-navigation.html [hunk 20]
```diff
@@ -12,7 +12,7 @@
           <button class="dropdown-toggle navbar-link" data-toggle="dropdown" id="lensPrimaryMenu"
               aria-controls="lensDropdownMenu">
             <span class="link-label">
-            <span class="sr-only">{% trans "Lens Menu." %}</span>
+            <span class="sr-only">{% trans "Navigation Menu." %}</span>
```
Comment (dsjen) [inline analytics_dashboard/templates/lens-navigation.html:15]: 👍

**Comment by downzer0** [comment]
👍

**Comment by dsjen** [comment]
🍰  🍰  🍰  🍰  🍰  🍰  🍰  🍰  🍰
